Adapted MemRegion::getDescriptiveName to handle ElementRegions

T-Gruber opened the pull request:
Fixes https://github.com/llvm/llvm-project/issues/84463

Changes:
- Adapted MemRegion::getDescriptiveName
- Added unittest to check name for a given clang::ento::ElementRegion
- Some format changes due to clang-format

Commit:
Adapted MemRegion::getDescriptiveName to handle ElementRegions

steakhal committed:
Revert unnecessary clang-format

steakhal committed:
Refine uses of auto

steakhal committed:
Fix license comment

steakhal committed:
Cleanup unittest includes

steakhal committed:
Rework testing

steakhal committed:
Rename unittest file

steakhal committed:
Cleanup redundant code

steakhal commented:
I hope you don't mind that I reworked the testing a bit, to be similar to the rest of the tests.
I also reverted unnecessary clang-formatting where it was not relevant.
I refined the `auto` usage, but other than that I left your code untouched.

As I did everything I thought, I'll invite others to review the PR.

T-Gruber commented:
Hi @steakhal, thank you very much for your additional changes!

steakhal committed:
clang-format

T-Gruber commented:
Hello @steakhal, I have just looked through the changes again. What is the advantage of using checkPreCall instead of checkLocation? I would very much appreciate some background information. Thanks for your help!

steakhal commented:
> Hello @steakhal, I have just looked through the changes again. What is the advantage of using checkPreCall instead of checkLocation? I would very much appreciate some background information. Thanks for your help!

Oh yes, I should have explained.
So, `checkLocation` can be triggered for all loads and store operations. Each time in the AST you have a `LValueToRValue` cast for reads, and others for write operations. This means that you can't really define a test where you have clear control of what region you wanted to dump.

Using `PreCall`, you can have your own API, that does what you think, exactly when it encounters that call.
This is similar to what we do in the `ExprInspection` checker, where we have analyzer debug intrinsics defined. Such as `clang_analyzer_dump(T)`. Search for it in the tests, and you will see what I'm talking about.

There is one more benefit of doing this way: There is already a communication channel for the stringified descriptive name: diagnostics.
One can easily capture and compare the diagnostics, like in the rest of the tests.

I was happy to help as I don't expect new contributors to be really involved with our specific testing.
To me what moved the needle was: well described expectation, well described and reproducible actual outcome. You even provided a fix, so I figured I'll help out with the rest.

Ask me if there is anything I left out.

T-Gruber commented:
Thanks again @steakhal! This is helpful and interesting background information. I really appreciate your help!

NagyDonat reviewed (changes requested):
I strongly suspect that there are some situations where this commit would behave incorrectly (assert on legal code or print incorrect output). See inline comment for explanation and a suggested solution (that I did not test).

steakhal commented on clang/lib/StaticAnalyzer/Core/MemRegion.cpp:
```diff
@@ -720,14 +720,20 @@ std::string MemRegion::getDescriptiveName(bool UseQuotes) const {
-      std::string Idx = ER->getDescriptiveName(false);
-      if (!Idx.empty()) {
-        ArrayIndices = (llvm::Twine("[") + Idx + "]" + ArrayIndices).str();
+    // Index is a SymbolVal.
+    else if (auto SI = ER->getIndex().getAs<nonloc::SymbolVal>()) {
+      if (SymbolRef SR = SI->getAsSymbol()) {
+        if (const MemRegion *OR = SR->getOriginRegion()) {
+          std::string Idx = OR->getDescriptiveName(false);
+          ArrayIndices = (llvm::Twine("[") + Idx + "]" + ArrayIndices).str();
+        }
       }
     }
+    // Index is neither a ConcreteInt nor SymbolVal, give up and return.
+    else {
+      assert(false && "we should have a descriptive name");
+      return "";
+    }
```
Let me elaborate what I think.
To set the scene, I'd like to point out that there was an unconditional infinite recursion in the previous code; and we in practice never seen crashes related to this so I think it's fair to say that currently we don't have any checkers exercising this path. This suggests to me that we could basically do anything here, that is better than a certain crash (aka. improve the previous handling).

So, at this point I'm okay to accept any code that at least sometimes work - given that we have a test for exercising that path.

And btw, if any of the sub-conditions fail, we hit the `else` branch and we assert in debug builds to get notified, and silently ignore the case (and may produce incomplete messages that people can still notice and report) in release builds.
Consequently, this shouldn't affect users in a blocking way.

T-Gruber replied:
Hi @NagyDonat, thank you for the hint. I improved the implementation based on your suggestion and also added 3 more test cases to the unittest.

steakhal commented:
> In addition to the above-mentioned issues there is also a conceptual problem with using `getOriginRegion()` to describe a symbol: it names the region where the symbol _originated_ (if it originated as the unknown initial value of a symbol), which is not necessarily the region where the symbol is stored _now_.
> 
> For example consider the (somewhat contrived) code
> 
> ```
> int matrix[10][10];
> int func(int x, int y) {
>   int tmp = x;
>   x = y;
>   y = tmp;
>   return matrix[x][-999]; 
> }
> ```
> 
> Here the checker alpha.security.ArrayBoundV2 (which uses `getDescriptiveName`) would report that
> 
> ```
> Access of 'matrix[y]' at negative byte offset
> ```
> 
> because the symbolic value of the index _originated from_ the parameter `y` (i.e. the symbol is identified as "this is the initial value of `y`).

I agree, but I don't see how this relates to this patch.
I think the original author of this code probably wanted to something what we propose here.
We can argue if that's a good thing or not, but that feels like a different topic to me.

Commit:
Extended else branch to handle symbolics in a better way & added corresponding unittest test cases

T-Gruber commented:
> In addition to the above-mentioned issues there is also a conceptual problem with using `getOriginRegion()` to describe a symbol: it names the region where the symbol _originated_ (if it originated as the unknown initial value of a symbol), which is not necessarily the region where the symbol is stored _now_.
> 
> For example consider the (somewhat contrived) code
> 
> ```
> int matrix[10][10];
> int func(int x, int y) {
>   int tmp = x;
>   x = y;
>   y = tmp;
>   return matrix[x][-999]; 
> }
> ```
> 
> Here the checker alpha.security.ArrayBoundV2 (which uses `getDescriptiveName`) would report that
> 
> ```
> Access of 'matrix[y]' at negative byte offset
> ```
> 
> because the symbolic value of the index _originated from_ the parameter `y` (i.e. the symbol is identified as "this is the initial value of `y`).

Yes, thats right. But the previous implementation ended up in an endless recursion if the index is not of type ConcreteInt. Therefore, the proposed version is a significant improvement (at least for my use cases).

NagyDonat reviewed (approved):
LGTM, thanks for adding the testcases!

I agree that the "sometimes reports the wrong variable" situation is a very significant improvement over the old "always infinite loop" situation; but this "convert a symbol to a variable name" step was completely missing from the old code, so I think it's relevant to mention that this commit implements it in a way that's not entirely correct.

I think it would be good to highlight this limitation with a FIXME comment (and perhaps a testcase), but I don't want to significantly delay the resolution of the infinite loop bug.

steakhal commented:
@T-Gruber Could you please add that test case with a FIXME?

T-Gruber commented:
Hi @steakhal, just to be sure. We are talking about a test case that covers that the index is "recognised" as a wrong variable (the matrix example above)?

steakhal commented:
> Hi @steakhal, just to be sure. We are talking about a test case that covers that the index is "recognised" as a wrong variable (the matrix example above)?

Exactly!

NagyDonat commented:
Since suggesting that "matrix" testcase I realized that it won't work, because `ArrayBoundV2` doesn't perform bounds checking for the lower-dimensional sub-arrays of a multidimensional array (it only checks that the accessed memory location is within the full array). For example in `int matrix[10][10];` it reports that `matrix[1][200]` overflows the bounds of "matrix", but doesn't report `matrix[1][50]`. I tried to eliminate this limitation, but that patch had to be reverted because the meaning of `ElementRegion` is ambiguous.

Due to this limitation, it's a bit difficult to ensure that `ArrayBoundV2` calls `getDescriptiveName()` on an `ElementRegion`, but I think adding a `FieldRegion` layer should work:
```
struct {
  int numbers[10];
} table[100];
int func(int x, int y) {
  int tmp = x;
  x = y;
  y = tmp;
  return table[x].numbers[-1]; 
}
```
Here I expect that the checker would (incorrectly) report something like `Access of 'table[y].numbers' at negative byte offset`.

Commit:
Added test case that leads to an incorrect index due to the use of getOriginRegion

T-Gruber commented:
Alright, I added a test case which leads to an incorrect index due to the use of getOriginRegion. I also added a FIXME comment to document that this case is not covered currently.

NagyDonat reviewed (approved):
LGTM.

Using this unit test framework is significantly cleaner than my suggestion that tries to work around the limitations of ArrayBoundV2.

Commit:
Adapted format

T-Gruber committed:
Merge branch 'main' into memregion_bug

steakhal committed:
Merge branch 'main' into memregion_bug
